compiler/src/dmd/statementsem.d

```diff
@@ -2758,6 +2758,8 @@ Statement statementSemanticVisit(Statement s, Scope* sc)
                 {
                     if (e0)
                         rs.error("expected return type of `%s`, not `%s`", tret.toChars(), resType.toChars());
+                    else if (tbret.isTypeNoreturn())
+                        rs.error("cannot return from `noreturn` function");
```

Comment:
A supplemental explaining you can use `while(1)` or assert 0 etc would be good - I've seen people assume noreturn acts top-down rather than requiring something to help the compiler.